### DISCUSSION TITLE
Should we merge testing binaries into a single `Tracker Client`?

### DISCUSSION THREAD

**Discussion by josecelano** [torrust/torrust-tracker#660] [Proposals]
Currently, we have some binaries that help testing trackers:

### HTTP Tracker Client

```console
cargo run --bin http_tracker_client announce http://127.0.0.1:7070 9c38422213e30bff212b30c360d26f9a02136422 | jq
cargo run --bin http_tracker_client scrape http://127.0.0.1:7070 9c38422213e30bff212b30c360d26f9a02136422 | jq
```

### UDP Tracker Client

```console
cargo run --bin udp_tracker_client announce 127.0.0.1:6969 9c38422213e30bff212b30c360d26f9a02136422 | jq
cargo run --bin udp_tracker_client scrape 127.0.0.1:6969 9c38422213e30bff212b30c360d26f9a02136422 | jq
```

### Tracker Checker

```console
cargo run --bin tracker_checker -- --config-path "./share/default/config/tracker_checker.json"
TORRUST_CHECKER_CONFIG_PATH="./share/default/config/tracker_checker.json" cargo run --bin tracker_checker
TORRUST_CHECKER_CONFIG=$(cat "./share/default/config/tracker_checker.json") cargo run --bin tracker_checker
```

I'm considering creating a single `Tracker Client`

### Tracker Client

```console
cargo run --bin tracker_client HTTP announce http://127.0.0.1:7070 9c38422213e30bff212b30c360d26f9a02136422 | jq
cargo run --bin tracker_client HTTP scrape http://127.0.0.1:7070 9c38422213e30bff212b30c360d26f9a02136422 | jq
cargo run --bin tracker_client UDP announce 127.0.0.1:6969 9c38422213e30bff212b30c360d26f9a02136422 | jq
cargo run --bin tracker_client UDP scrape 127.0.0.1:6969 9c38422213e30bff212b30c360d26f9a02136422 | jq
cargo run --bin tracker_client CHECK -- --config-path "./share/default/config/tracker_checker.json"
TORRUST_CHECKER_CONFIG_PATH="./share/default/config/tracker_checker.json" cargo run --bin tracker_client CHECK
TORRUST_CHECKER_CONFIG=$(cat "./share/default/config/tracker_checker.json") cargo run --bin tracker_client CHECK
```

The client will have subcommands for HTTP trackers, UDP trackers, and the Tracker Checker.

What do you think @WarmBeer @da2ce7?

**Comment by da2ce7** [comment]
I think the protocol should be detected from the url, and the command should be what action.

**Reply by da2ce7** [reply]
Also, we can have an argument to force a particular protocol, making it an error if the url is in the wrong format.

**Reply by josecelano** [reply]
> Also, we can have an argument to force a particular protocol, making it an error if the url is in the wrong format.

Hi @da2ce7 , I was planning to use subcommands to keep the code organized for UDP and HTTP. That makes code simpler but maybe UX less friendly.

**Reply by josecelano** [reply]
And @da2ce7 I've been using JSON as the default output format for the UDP and HTTP clients, but a human-friendly output for the Checker:

![image](https://github.com/torrust/torrust-tracker/assets/58816/ab78bdcd-f885-495e-8a07-1ebcd93b1449)

Maybe we should make the human-friendly format the default one for all commands and add an option `--format=json` when you want the output in JSON. We can make this change in the next [milestone](https://github.com/torrust/torrust-tracker/milestone/2).

**Reply by josecelano** [reply]
[I'm still working on the Checker](https://github.com/torrust/torrust-tracker/issues/637). Maybe I can change it to use JSON. It should not take more than an hour.

**Reply by josecelano** [reply]
@da2ce7 we could also add a subcommand for the [Tracker API](https://github.com/torrust/torrust-tracker/blob/develop/tests/servers/api/v1/client.rs) to run commands like this:

```console
cargo run --bin tracker_checker API -- --config-path "./tracker_client.json" torrents
cargo run --bin tracker_checker API -- --config-path "./tracker_client.json" torrent 9c38422213e30bff212b30c360d26f9a02136422
```

But I would like the Tracker Client to be generic and the API is specific for Turrust Tracker. Maybe in that case it would be better to have a Tracker API Console Client.

**Comment by josecelano** [comment]
[Clients were extracted into a new package](https://github.com/torrust/torrust-tracker/issues/1067), `bittorrent-tracker-client`
